File: org.kevoree.modeling.cpp.microframework/src/main/java/org/kevoree/modeling/api/trace/ModelTraceApplicator.kt
package org.kevoree.modeling.api.trace

import org.kevoree.modeling.api.util.ActionType
import org.kevoree.modeling.api.KMFContainer
import org.kevoree.modeling.api.KMFFactory

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 06/08/13
 * Time: 08:54
 */
class ModelTraceApplicator(val targetModel: KMFContainer, val factory: KMFFactory) {

    var pendingObj: KMFContainer? = null
    var pendingParent: KMFContainer? = null
    var pendingParentRefName: String? = null
    var pendingObjPath: String? = null

    private fun tryClosePending(srcPath: String?) {
        if(pendingObj != null && pendingObjPath != srcPath){
            pendingParent!!.reflexiveMutator(ActionType.ADD, pendingParentRefName!!, pendingObj,true,true)
            pendingObj = null
            pendingObjPath = null
            pendingParentRefName = null
            pendingParent = null
        }
    }

    public fun createOrAdd(previousPath: String?, target: KMFContainer, refName: String, potentialTypeName: String?) {
        val targetElem: Any? = if(previousPath != null){
            targetModel.findByPath(previousPath)
        } else {
            null
        }
        if(targetElem != null){
            target.reflexiveMutator(ActionType.ADD, refName, targetElem,true,true)
        } else {
            //add to pending
            pendingObj = factory.create(potentialTypeName!!)
            pendingObjPath = previousPath;
            pendingParentRefName = refName;
            pendingParent = target;
        }
    }

    public fun applyTraceOnModel(traceSeq: TraceSequence) {

        for(trace in traceSeq.traces){
            var target: KMFContainer = targetModel;
            if(trace is ModelAddTrace){
                val castedTrace = trace as ModelAddTrace
                tryClosePending(null);
                if(trace.srcPath != ""){
                    target = targetModel.findByPath(castedTrace.srcPath) as KMFContainer;
                }
                createOrAdd(castedTrace.previousPath, target, castedTrace.refName, castedTrace.typeName)
            }
            if(trace is ModelAddAllTrace){
                val castedTrace = trace as ModelAddAllTrace
                tryClosePending(null);
                var i = 0
                for(path in castedTrace.previousPath!!){
                    createOrAdd(path, target, castedTrace.refName, castedTrace.typeName!!.get(i))
                    i++
                }
            }
            if(trace is ModelRemoveTrace){
                val castedTrace = trace as ModelRemoveTrace
                tryClosePending(trace.srcPath);
                var tempTarget: KMFContainer? = targetModel;
                if(trace.srcPath != ""){
                    tempTarget = targetModel.findByPath(castedTrace.srcPath) as? KMFContainer;
                }
                if(tempTarget != null){
                    //Potentially null if top tree already dropped
                    tempTarget!!.reflexiveMutator(ActionType.REMOVE, castedTrace.refName, targetModel.findByPath(castedTrace.objPath),true,true)
                }
            }
            if(trace is ModelRemoveAllTrace){
                val castedTrace = trace as ModelRemoveAllTrace
                tryClosePending(trace.srcPath);
                var tempTarget: KMFContainer? = targetModel;
                if(trace.srcPath != ""){
                    tempTarget = targetModel.findByPath(castedTrace.srcPath) as? KMFContainer;
                }
                if(tempTarget != null){
                    tempTarget!!.reflexiveMutator(ActionType.REMOVE_ALL, castedTrace.refName, null,true,true)
                }
            }
            if(trace is ModelSetTrace){
                val castedTrace = trace as ModelSetTrace
                tryClosePending(trace.srcPath);
                if(trace.srcPath != "" && castedTrace.srcPath != pendingObjPath){
                    var tempObject = targetModel.findByPath(castedTrace.srcPath)
                    if(tempObject == null){

                        throw Exception("Set Trace source not found for path : " + castedTrace.srcPath + "/ pending " + pendingObjPath + "\n" + trace.toString())
                    }
                    target = tempObject as KMFContainer;
                } else {
                    if(castedTrace.srcPath == pendingObjPath && pendingObj != null){
                        target = pendingObj!!;
                    }
                }
                if(castedTrace.content != null){
                    target.reflexiveMutator(ActionType.SET, castedTrace.refName, castedTrace.content,true,true)
                } else {
                    var targetContentPath: Any? = if(castedTrace.objPath != null){
                        targetModel.findByPath(castedTrace.objPath!!)
                    } else {
                        null
                    };
                    if(targetContentPath != null){
                        target.reflexiveMutator(ActionType.SET, castedTrace.refName, targetContentPath,true,true)
                    } else {
                        if(castedTrace.typeName != null && castedTrace.typeName != ""){
                            createOrAdd(castedTrace.objPath, target, castedTrace.refName, castedTrace.typeName) //must create the pending element
                        } else {
                            target.reflexiveMutator(ActionType.SET, castedTrace.refName, targetContentPath,true,true) //case real null content
                        }
                    }
                }
            }
        }
        tryClosePending(null);
    }

}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/trace/ModelTraceApplicator.ts
///<reference path="../Callback.ts"/>
///<reference path="../KActionType.ts"/>
///<reference path="../KObject.ts"/>
///<reference path="../meta/Meta.ts"/>
///<reference path="../meta/MetaAttribute.ts"/>
///<reference path="../meta/MetaClass.ts"/>
///<reference path="../meta/MetaReference.ts"/>
///<reference path="../util/CallBackChain.ts"/>
///<reference path="../util/Helper.ts"/>

class ModelTraceApplicator {

  private targetModel: KObject<any,any> = null;
  private pendingObj: KObject<any,any> = null;
  private pendingParent: KObject<any,any> = null;
  private pendingParentRef: MetaReference = null;
  private pendingObjKID: number = null;

  constructor(targetModel: KObject<any,any>) {
    this.targetModel = targetModel;
  }

  private tryClosePending(srcKID: number): void {
    if (this.pendingObj != null && !(this.pendingObjKID.equals(srcKID))) {
      this.pendingParent.mutate(KActionType.ADD, this.pendingParentRef, this.pendingObj, true);
      this.pendingObj = null;
      this.pendingObjKID = null;
      this.pendingParentRef = null;
      this.pendingParent = null;
    }
  }

  public createOrAdd(previousPath: number, target: KObject<any,any>, reference: MetaReference, metaClass: MetaClass, callback: Callback<Throwable>): void {
    if (previousPath != null) {
      this.targetModel.view().lookup(previousPath, {on:function(targetElem: KObject){
      if (targetElem != null) {
        target.mutate(KActionType.ADD, reference, targetElem, true);
        callback.on(null);
      } else {
        if (metaClass == null) {
          callback.on(new Exception("Unknow typeName for potential path " + previousPath + ", to store in " + reference.metaName() + ", unconsistency error"));
        } else {
          this.pendingObj = this.targetModel.view().createFQN(metaClass.metaName());
          this.pendingObjKID = previousPath;
          this.pendingParentRef = reference;
          this.pendingParent = target;
          callback.on(null);
        }
      }
}});
    } else {
      if (metaClass == null) {
        callback.on(new Exception("Unknow typeName for potential path " + previousPath + ", to store in " + reference.metaName() + ", unconsistency error"));
      } else {
        this.pendingObj = this.targetModel.view().createFQN(metaClass.metaName());
        this.pendingObjKID = previousPath;
        this.pendingParentRef = reference;
        this.pendingParent = target;
        callback.on(null);
      }
    }
  }

  public applyTraceSequence(traceSeq: TraceSequence, callback: Callback<Throwable>): void {
    Helper.forall(traceSeq.traces(), {on:function(modelTrace: ModelTrace, next: Callback<Throwable>){
    this.applyTrace(modelTrace, next);
}}, {on:function(throwable: Throwable){
    if (throwable != null) {
      callback.on(throwable);
    } else {
      this.tryClosePending(null);
      callback.on(null);
    }
}});
  }

  public applyTrace(trace: ModelTrace, callback: Callback<Throwable>): void {
    if (trace instanceof ModelAddTrace) {
      var addTrace: ModelAddTrace = <ModelAddTrace>trace;
      this.tryClosePending(null);
      this.targetModel.view().lookup(trace.getSrcKID(), {on:function(resolvedTarget: KObject){
      if (resolvedTarget == null) {
        callback.on(new Exception("Add Trace source not found for path : " + trace.getSrcKID() + " pending " + this.pendingObjKID + "\n" + trace.toString()));
      } else {
        this.createOrAdd(addTrace.getPreviousKID(), resolvedTarget, <MetaReference>trace.getMeta(), addTrace.getMetaClass(), callback);
      }
}});
    } else {
      if (trace instanceof ModelRemoveTrace) {
        var removeTrace: ModelRemoveTrace = <ModelRemoveTrace>trace;
        this.tryClosePending(trace.getSrcKID());
        this.targetModel.view().lookup(trace.getSrcKID(), {on:function(targetElem: KObject){
        if (targetElem != null) {
          this.targetModel.view().lookup(removeTrace.getObjKID(), {on:function(remoteObj: KObject){
          targetElem.mutate(KActionType.REMOVE, <MetaReference>trace.getMeta(), remoteObj, true);
          callback.on(null);
}});
        } else {
          callback.on(null);
        }
}});
      } else {
        if (trace instanceof ModelSetTrace) {
          var setTrace: ModelSetTrace = <ModelSetTrace>trace;
          this.tryClosePending(trace.getSrcKID());
          if (!trace.getSrcKID().equals(this.pendingObjKID)) {
            this.targetModel.view().lookup(trace.getSrcKID(), {on:function(tempObject: KObject){
            if (tempObject == null) {
              callback.on(new Exception("Set Trace source not found for path : " + trace.getSrcKID() + " pending " + this.pendingObjKID + "\n" + trace.toString()));
            } else {
              tempObject.set(<MetaAttribute>setTrace.getMeta(), setTrace.getContent());
              callback.on(null);
            }
}});
          } else {
            if (this.pendingObj == null) {
              callback.on(new Exception("Set Trace source not found for path : " + trace.getSrcKID() + " pending " + this.pendingObjKID + "\n" + trace.toString()));
            } else {
              this.pendingObj.set(<MetaAttribute>setTrace.getMeta(), setTrace.getContent());
              callback.on(null);
            }
          }
        } else {
          callback.on(new Exception("Unknow trace " + trace));
        }
      }
    }
  }

}

